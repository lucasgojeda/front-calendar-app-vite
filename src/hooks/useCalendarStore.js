import { useDispatch, useSelector } from 'react-redux';
import Swal from "sweetalert2";

import calendarApi from '../api/calendarApi';

import { fetchConToken } from "../helpers/fetch"

import {
    eventAddNew,
    eventClearActiveEvent,
    eventDeleted,
    eventLoaded,
    eventLogout,
    eventSetActive,
    eventUpdated
} from "../store/slices/calendarSlice";


/**
 * En este hook encontramos todas las acciones y propiedades relacionados con los eventos.
 * @module useCalendarStore
 * @example evento = {
 *     title: 'Cumpleaños del jefe',
 *     start: Date,
 *     end: Date,
 *     notes: 'Comprar el pastel',
 *     user: {
 *         _id: '123',
 *         name: 'Fernando'
 *     }
 * }
 */
export const useCalendarStore = () => {

    const dispatch = useDispatch();
    const { events, activeEvent } = useSelector(state => state.calendar);
    const { name, uid } = useSelector(state => state.auth);


    /**
 * Esta función se encarga de crear un nuevo evento.
 * @function
 * @async
 * @param {Object} event - Evento a ser creado.
 */
    const eventStartAddNew = async (event) => {

        try {

            event.user = {
                uid,
                name
            }

            const { data: { msg } } = await calendarApi.post('events', event);

            if (msg._id !== undefined) {

                dispatch(eventAddNew(msg));
            }

        } catch (error) {
            console.log(error);
        }
    }

    /**
     * Esta función se encarga de actualizar un evento.
     * @function
     * @async
     * @param {Object} event - Evento a ser actualizado.
     */
    const eventStartUpdated = async (event) => {

        try {

            const { data: { msg } } = await calendarApi.put(`events/${event._id}`, { event, uid });

            if (msg.user.uid) {
                dispatch(eventUpdated({ ...msg }));
                dispatch(eventStartLoaded());
            } else {
                Swal.fire('Error', msg, 'error');
            }

        } catch (error) {
            console.log(error);
        }
    }

    /**
     * Esta función se encarga de cargar todas los eventos.
     * @function
     * @async
     */
    const eventStartLoaded = async () => {

        try {

            const { data: { eventos } } = await calendarApi.get('events');

            if (eventos) {

                dispatch(eventLoaded(eventos))

            }

        } catch (error) {
            console.log(error);
        }
    }

    /**
     * Esta función se encarga de eliminar un evento.
     * @function
     * @async
     * @param {Object} event - Evento a ser eliminado.
     */
    const eventStartDeleted = async (event) => {

        try {

            const resp = await fetchConToken(`events/${event._id}`, { uid }, 'DELETE');
            const body = await resp.json();

            if (body.msg === "Todo ok") {
                dispatch(eventDeleted({ ...event }));
                dispatch(eventStartLoaded());
            } else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);
        }
    }


    const startEventClearActiveEvent = () => {

        dispatch(eventClearActiveEvent());
    }

    const startEventSetActive = (event) => {

        dispatch(eventSetActive(event));
    }

    const startEventLogout = (event) => {

        dispatch(eventLogout());
    }




    return {
        //* Propiedades
        events,
        activeEvent,

        //* Métodos
        eventStartAddNew,
        eventStartUpdated,
        eventStartLoaded,
        eventStartDeleted,
        startEventClearActiveEvent,
        startEventSetActive,
        startEventLogout
    }
}
