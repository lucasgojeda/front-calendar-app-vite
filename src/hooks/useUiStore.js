import { useDispatch, useSelector } from 'react-redux';
import { uiCloseModal, uiOpenModal } from '../store/slices/uiSlice';


export const useUiStore = () => {

    const { modalOpen } = useSelector( state => state.ui );
    const dispatch = useDispatch();

    const startUiOpenModal = () => {
        dispatch( uiOpenModal() )
    }

    const startUiCloseModal = () => {
        dispatch( uiCloseModal() )
    }


    return {
        //* Propiedades
        modalOpen,

        //* MÃ©todos
        startUiOpenModal,
        startUiCloseModal,
    }

}