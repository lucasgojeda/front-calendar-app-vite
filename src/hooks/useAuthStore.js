import { useDispatch, useSelector } from 'react-redux';
import Swal from 'sweetalert2';

import calendarApi from '../api/calendarApi';
import { getEnvironmets } from '../helpers/getEnvironmets';

import {
    authCheckingFinish,
    authLogin,
    authLogout
} from '../store/slices/authSlice';

import { useCalendarStore } from './useCalendarStore';

/**
 * En este hook encontramos todas las acciones y propiedades relacionados a 
 * la autenticación.
 * @module useAuthStore
 */
export const useAuthStore = () => {

    const dispatch = useDispatch();
    const { checking, uid, name } = useSelector(state => state.auth);
    const { eventStartLoaded } = useCalendarStore();

    /**
     * Esta función realiza el inicio de sesión.
     * @function
     * @async
     * @param {String} correo - Correo del usuario.
     * @param {String} password - Contraseña del usuario.
     */
    const StartLogin = async (correo, password) => {

        try {

            const { data: body } = await calendarApi.post('auth/login', { correo, password });

            if (body.token) {

                localStorage.setItem('token', body.token);
                localStorage.setItem('token-init-date', new Date().getTime());

                dispatch(authLogin({
                    uid: body.user._id,
                    name: body.user.name
                }));

                dispatch(eventStartLoaded());
            } else {
                console.log(body.msg);
                return Swal.fire('Error', body.msg, 'error');
            }


        } catch (error) {
            if (getEnvironmets().VITE_MODE !== 'test') {

                console.log(error);
                Swal.fire('Error', error, 'error');

            }
        }
    }

    /**
     * Esta función realiza el registro de un nuevo usuario.
     * @function
     * @async
     * @param {String} name - Nombre del nuevo usuario.
     * @param {String} correo - Correo del nuevo usuario.
     * @param {String} password - Contraseña del nuevo usuario.
     * @param {String} role - Role del nuevo usuario (usuario por defecto).
     */
    const startRegister = async (name, correo, password, role = "ADMIN_ROLE") => {

        try {

            const { data: body } = await calendarApi.post('users', { name, correo, password, role });

            if (body.token) {

                localStorage.setItem('token', body.token);
                localStorage.setItem('token-init-date', new Date().getTime());

                dispatch(authLogin({
                    uid: body.user._id,
                    name: body.user.name
                }));
            } else {
                console.log(body.errors);
                return Swal.fire('Error', body.errors[0].msg, 'error');
            }


        } catch (error) {
            Swal.fire('Error', error, 'error');
            console.log(error);
        }
    }

    /**
     * Esta función realiza el chequeo de si un usuario está logueado mediante la 
     * autenticación por token cada vez que recarga la pagina.
     * @function
     * @async
     */
    const startChecking = async () => {

        const token = localStorage.getItem('token');
        if (!token) return dispatch(authCheckingFinish());

        try {

            const { data } = await calendarApi.get('auth/renew');

            if (data !== undefined) {

                localStorage.setItem('token', data.token);
                localStorage.setItem('token-init-date', new Date().getTime());

                dispatch(authLogin({
                    uid: data.uid,
                    name: data.name
                }));
            } else {
                dispatch(authCheckingFinish());
            }

        } catch (error) {
            dispatch(authCheckingFinish());
            console.log(error);
        }
    }

    /**
     * Esta función realiza el cierre de sesión en la aplicación limpiando el localStorage 
     * y los datos del usuario del store en redux.
     * @function
     */
    const startLogout = async () => {

        localStorage.clear();

        dispatch(authLogout())
    }


    return {
        //* Propiedades
        checking,
        uid,
        name,

        //* Métodos
        StartLogin,
        startRegister,
        startChecking,
        startLogout,
    }

}