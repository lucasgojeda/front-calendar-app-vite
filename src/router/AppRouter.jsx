import React, { useEffect } from 'react';
import {
    Routes,
    Route,
    BrowserRouter,
    Navigate
} from "react-router-dom";


import { LoginScreen } from '../components/auth/LoginScreen';
import { CalendarScreen } from '../components/calendar/CalendarScreen';

import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

import { useAuthStore } from '../hooks/useAuthStore';

/**
 * Este componente maneja las rutas "/login" y "/*", que coresponden a 
 * los componentes LoginScreen y CalendarScreen.
 * @module AppRouter
 */
export const AppRouter = () => {

    const { checking, uid, startChecking } = useAuthStore();

    /** 
     * El siguiente useEffect se encarga de realizar el chequeo del usuario cada vez que 
     * se renderiza este componente.
     */
    useEffect(() => {

        startChecking();

    }, []);

    /**
     * Si el estado del inicio de sesión del usuario aún no ha sido verificado, entonces 
     * se mostrará solo el siguiente h5.
     */
    if (checking) {
        return <h5>Espere...</h5>
    }

    return (
            <Routes>

                <Route path="/login" element={
                    <PublicRoute isAutenticated={!!uid}>
                        <LoginScreen />
                    </PublicRoute>
                } />

                <Route path="/" element={
                    <PrivateRoute isAutenticated={!!uid}>
                        <CalendarScreen />
                    </PrivateRoute>
                } />

                <Route path="/*" element={<Navigate to="/" />} />

            </Routes>
    );
};