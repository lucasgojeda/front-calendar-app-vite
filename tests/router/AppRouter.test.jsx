import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import { useAuthStore } from "../../src/hooks/useAuthStore";
import { AppRouter } from "../../src/router/AppRouter";

jest.mock('../../src/hooks/useAuthStore');

jest.mock('../../src/components/calendar/CalendarScreen', () => ({
    CalendarScreen: () => <h1>CalendarScreen</h1>
}));


describe('Pruebas en el componente AppRouter', () => {

    const mockStartChecking = jest.fn();

    beforeEach(() => jest.clearAllMocks());

    test('debe de mostrar la pantalla de carga y llamar startChecking', () => {

        useAuthStore.mockReturnValue({
            checking: true,
            uid: '123',
            startChecking: mockStartChecking
        })

        render(
            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>);

        const title = screen.getByText('Espere...');

        expect(title).toBeTruthy();
        expect(mockStartChecking).toHaveBeenCalled();

    });

    test('debe de mostrar el login en caso de no estar autenticado', () => {

        useAuthStore.mockReturnValue({
            checking: false,
            uid: null,
            name: null,
            startChecking: mockStartChecking
        })

        const { container } = render(
            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>);

        expect(screen.getByText('Ingreso')).toBeTruthy();
        expect(container).toMatchSnapshot();
    });

    test('debe de mostrar el calendario en caso de estar autenticado', () => {

        useAuthStore.mockReturnValue({
            checking: false,
            uid: '123',
            name: 'Lucas Ojeda',
            startChecking: mockStartChecking
        })

        const { container } = render(
            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>);


        expect(screen.getByText('CalendarScreen')).toBeTruthy();
    });
});