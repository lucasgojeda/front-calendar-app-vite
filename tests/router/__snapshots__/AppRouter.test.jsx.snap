// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente AppRouter debe de mostrar el login en caso de no estar autenticado 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              name="lCorreo"
              placeholder="Correo"
              type="text"
              value="test20@test.com"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              name="lPassword"
              placeholder="Contraseña"
              type="password"
              value="123456"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              name="rName"
              placeholder="Nombre"
              type="text"
              value="test"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              name="rCorreo"
              placeholder="Correo"
              type="email"
              value="test20@test.com"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              name="rPassword"
              placeholder="Contraseña"
              type="password"
              value="123456"
            />
          </div>
          <div
            class="form-group"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
